Framework/Algorithms/src/GetEi2.cpp

```diff
@@ -276,31 +277,33 @@ double GetEi2::calculateEi(const double initial_guess) {
   }
 }
 
-/** Gets the distance between the source and detectors whose workspace index is
+/**
+ * Gets the distance between the source and detectors whose workspace index is
  * passed
  *  @param ws_index :: The workspace index of the detector
+ *  @param spectrumInfo :: A spectrum info object for the input workspace
  *  @return The distance between the source and the given detector(or
  * DetectorGroup)
  *  @throw runtime_error if there is a problem
  */
-double GetEi2::getDistanceFromSource(size_t ws_index) const {
+double GetEi2::getDistanceFromSource(size_t ws_index,
+                                     const SpectrumInfo &spectrumInfo) const {
   g_log.debug() << "Computing distance between spectrum at index '" << ws_index
                 << "' and the source\n";
 
   const IComponent_const_sptr source = m_input_ws->getInstrument()->getSource();
-  // Retrieve a pointer detector
-  IDetector_const_sptr det = m_input_ws->getDetector(ws_index);
-  if (!det) {
+  if (!spectrumInfo.hasDetectors(ws_index)) {
     std::ostringstream msg;
     msg << "A detector for monitor at workspace index " << ws_index
         << " cannot be found. ";
     throw std::runtime_error(msg.str());
   }
   if (g_log.is(Logger::Priority::PRIO_DEBUG)) {
-    g_log.debug() << "Detector position = " << det->getPos()
+    g_log.debug() << "Detector position = " << spectrumInfo.position(ws_index)
                   << ", Source position = " << source->getPos() << "\n";
   }
-  const double dist = det->getDistance(*source);
+  const double dist =
+      spectrumInfo.position(ws_index).distance(source->getPos());
```

Comment:
A similar thing happens for theta, and I actually have a test with an explicit comment for this: https://github.com/mantidproject/mantid/blob/master/Framework/API/test/SpectrumInfoTest.h#L170

This is the way it is done in `DetectorGroup`, and I think it is correct, since 2-theta is the *scattering* angle, so the result should be the average of the scattering angles. If asked for the position of a `DetectorGroup` in polar coordinates, the result should be the angle of the average position.

Likewise, we are defining L2 as the average L2.

Defining distance for a group as the average distance does not seem to make sense in general, in my opinion, so maybe this here is a special case?